storage/buffer.go

```diff
@@ -161,12 +161,19 @@ func (s *dbBuffer) bucketState(
 	return shouldRead, needsDrain, needsReset
 }
 
-func (s *dbBuffer) DrainAndReset() {
+func (s *dbBuffer) DrainAndReset(forced bool) {
 	now := s.nowFn()
 	s.forEachBucketAsc(now, func(b *dbBufferBucket, current time.Time) {
-		_, needsDrain, needsReset := s.bucketState(now, b, current)
-		if !needsDrain && !needsReset {
-			// No action necessary
+		needsDrain, needsReset := true, true
+		if !forced {
+			_, needsDrain, needsReset = s.bucketState(now, b, current)
+			if !needsDrain && !needsReset {
+				// No action necessary
+				return
+			}
+		} else if b.drained || b.lastWriteAt.IsZero() {
+			// No action necessary if the bucket has been drained,
+			// or there is no data in the bucket.
```

Comment:
yeah i guess it doesn't hurt to drain/reset when it's forced and the first if condition is not satisfied. Will update.